File: 大三上-第5学期/编译原理与技术/编译原理汇报/code/Python/unify.py
"""
unify.py - 类型合一（Unification）
Robinson 统一算法的 Python 实现
"""

from typing import Tuple, Optional
from hm_types import (
    Type, TVar, TInt, TBool, TFun, TPair, 
    Subst
)


class UnificationError(Exception):
    """合一失败异常"""
    pass


def occurs_check(tvar_id: int, t: Type) -> bool:
    """
    出现检查（Occurs Check）
    检查类型变量是否出现在类型 t 中
    用于防止产生无限类型，如 α = α -> β
    """
    if isinstance(t, TVar):
        return t.id == tvar_id
    elif isinstance(t, (TInt, TBool)):
        return False
    elif isinstance(t, TFun):
        return occurs_check(tvar_id, t.param) or occurs_check(tvar_id, t.result)
    elif isinstance(t, TPair):
        return occurs_check(tvar_id, t.fst) or occurs_check(tvar_id, t.snd)
    return False


def unify(t1: Type, t2: Type) -> Subst:
    """
    合一两个类型
    
    返回最一般合一子（MGU）S，使得 S(t1) = S(t2)
    如果无法合一，抛出 UnificationError
    
    合一规则：
    1. 相同类型 → 空替换
    2. 类型变量 α 与 τ → [α := τ]（需要出现检查）
    3. τ1 -> τ2 与 σ1 -> σ2 → 递归合一
    4. (τ1, τ2) 与 (σ1, σ2) → 递归合一
    5. 否则 → 失败
    """
    # 1. 相同类型
    if t1 == t2:
        return Subst.empty()
    
    # 2a. t1 是类型变量
    if isinstance(t1, TVar):
        if occurs_check(t1.id, t2):
            raise UnificationError(
                f"Occurs check failed: {t1} occurs in {t2} (infinite type)"
            )
        return Subst.single(t1.id, t2)
    
    # 2b. t2 是类型变量
    if isinstance(t2, TVar):
        if occurs_check(t2.id, t1):
            raise UnificationError(
                f"Occurs check failed: {t2} occurs in {t1} (infinite type)"
            )
        return Subst.single(t2.id, t1)
    
    # 3. 两个函数类型
    if isinstance(t1, TFun) and isinstance(t2, TFun):
        # 先合一参数类型
        s1 = unify(t1.param, t2.param)
        # 将 s1 应用到返回类型，再合一
        s2 = unify(
            t1.result.apply_subst(s1),
            t2.result.apply_subst(s1)
        )
        # 组合替换
        return s2.compose(s1)
    
    # 4. 两个二元组类型
    if isinstance(t1, TPair) and isinstance(t2, TPair):
        s1 = unify(t1.fst, t2.fst)
        s2 = unify(
            t1.snd.apply_subst(s1),
            t2.snd.apply_subst(s1)
        )
        return s2.compose(s1)
    
    # 5. 类型不匹配
    raise UnificationError(
        f"Type mismatch: cannot unify {t1} with {t2}"
    )

File: 大三上-第5学期/编译原理与技术/编译原理汇报/code/C/src/unify.c
/**
 * unify.c - 类型合一（Unification）实现
 * Robinson 统一算法
 */

#define _GNU_SOURCE
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include "../include/types.h"
#include "../include/type_utils.h"
#include "../include/subst.h"
#include "../include/unify.h"

/**
 * 出现检查（Occurs Check）
 * 防止产生无限类型
 */
bool occurs_check(TVar tv, Type *t) {
    if (t == NULL) return false;
    
    switch (t->kind) {
        case T_VAR:
            return tvar_equal(tv, t->tvar);
        case T_INT:
        case T_BOOL:
            return false;
        case T_FUN:
        case T_PAIR:
            return occurs_check(tv, t->left) || occurs_check(tv, t->right);
    }
    return false;
}

static UnifyResult make_success(Subst s) {
    UnifyResult r = { .success = true, .subst = s, .error_msg = NULL };
    return r;
}

static UnifyResult make_error(const char *msg) {
    UnifyResult r = {
        .success = false,
        .subst = subst_empty(),
        .error_msg = strdup(msg)
    };
    return r;
}

/**
 * 合一算法核心实现
 * unify(τ1, τ2) 返回最一般合一子 S 使得 S(τ1) = S(τ2)
 */
UnifyResult unify(Type *t1, Type *t2) {
    // 两个类型相同
    if (type_equal(t1, t2)) {
        return make_success(subst_empty());
    }
    
    // t1 是类型变量
    if (t1->kind == T_VAR) {
        if (occurs_check(t1->tvar, t2)) {
            return make_error("Occurs check failed: infinite type");
        }
        return make_success(subst_single(t1->tvar, t2));
    }
    
    // t2 是类型变量
    if (t2->kind == T_VAR) {
        if (occurs_check(t2->tvar, t1)) {
            return make_error("Occurs check failed: infinite type");
        }
        return make_success(subst_single(t2->tvar, t1));
    }
    
    // 两个函数类型
    if (t1->kind == T_FUN && t2->kind == T_FUN) {
        // 先合一参数类型
        UnifyResult r1 = unify(t1->left, t2->left);
        if (!r1.success) return r1;
        
        // 将替换应用到返回类型
        Type *right1 = apply_subst_type(&r1.subst, t1->right);
        Type *right2 = apply_subst_type(&r1.subst, t2->right);
        
        // 合一返回类型
        UnifyResult r2 = unify(right1, right2);
        free_type(right1);
        free_type(right2);
        
        if (!r2.success) {
            // 清理 r1 的内存
            return r2;
        }
        
        // 组合两个替换
        Subst composed = subst_compose(r2.subst, r1.subst);
        return make_success(composed);
    }
    
    // 两个二元组类型
    if (t1->kind == T_PAIR && t2->kind == T_PAIR) {
        UnifyResult r1 = unify(t1->left, t2->left);
        if (!r1.success) return r1;
        
        Type *right1 = apply_subst_type(&r1.subst, t1->right);
        Type *right2 = apply_subst_type(&r1.subst, t2->right);
        
        UnifyResult r2 = unify(right1, right2);
        free_type(right1);
        free_type(right2);
        
        if (!r2.success) return r2;
        
        Subst composed = subst_compose(r2.subst, r1.subst);
        return make_success(composed);
    }
    
    // 类型不匹配
    char error_buf[256];
    snprintf(error_buf, sizeof(error_buf),
             "Type mismatch: cannot unify %s with %s",
             type_to_string(t1), type_to_string(t2));
    return make_error(error_buf);
}

